### PR TITLE
Metrics Feature

### PR DESCRIPTION
**Pull request by AnujaLK**

### CODE REVIEW

File: modules/andes-core/broker/src/main/java/org/wso2/andes/metrics/MetricsConstants.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+ */
+
+package org.wso2.andes.metrics;
+
+/**
+ * Will maintain the list of metrics
+ */
+public class MetricsConstants {
+
+    public static final String PREFIX = "org.wso2.mb.";
```
Comment (pumudu88) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/metrics/MetricsConstants.java]: add a comment

Comment (AnujaLK) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/metrics/MetricsConstants.java]: Fixed with PR - https://github.com/wso2/andes/pull/202

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/distruptor/delivery/DeliveryEventHandler.java [hunk 2]
```diff
@@ -87,6 +90,11 @@ public void onEvent(DeliveryEventData deliveryEventData, long sequence, boolean
                 }
                 if (subscription.isActive()) {
                     subscription.sendMessageToSubscriber(message, deliveryEventData.getAndesContent());
+
+                    //Adding metrics meter for ack rate
+                    Meter msgMeter = MetricManager.meter(Level.INFO, this.getClass() + MetricsConstants.MSG_SENT_RATE);
+                    msgMeter.mark();
+
                 } else {
```
Comment (abeykoon) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/distruptor/delivery/DeliveryEventHandler.java]: Better break by "+" line and keep it at the beginning of next line.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/distruptor/inbound/MessagePreProcessor.java [hunk 3]
```diff
@@ -198,6 +203,11 @@ private void handleTopicRoutine(InboundEventContainer event, AndesMessage messag
                 // Even though we drop the message pub ack needs to be sent
                 event.pubAckHandler.ack(message.getMetadata());
 
+                //Adding metrics meter for ack rate
+                Meter ackMeter = MetricManager.meter(Level.INFO, this.getClass() + MetricsConstants.ACK_SENT_RATE);
+                ackMeter.mark();
+
```
Comment (abeykoon) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/distruptor/inbound/MessagePreProcessor.java]: Let us break by "+"

File: modules/andes-core/broker/src/main/java/org/wso2/andes/store/rdbms/RDBMSMessageStoreImpl.java [hunk 4]
```diff
@@ -647,8 +655,8 @@ private void addMetadataToBatch(PreparedStatement preparedStatement,
                                     final String queueName) throws SQLException {
 
 
-        Context context = DataAccessMatrixManager.addAndGetTimer(MatrixConstants.ADD_META_DATA_TO_BATCH, this).start();
-
+        Context context = MetricManager.timer(Level.INFO, MetricsConstants.ADD_META_DATA_TO_BATCH).start();
+        Context contextWrite = MetricManager.timer(Level.INFO, MetricsConstants.DB_WRITE).start();
```
Comment (sdkottegoda) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/store/rdbms/RDBMSMessageStoreImpl.java]: There isn't a DB call here. Do we need to start a contextWrite?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionStore.java [hunk 5]
```diff
@@ -748,4 +756,33 @@ public void deactivateClusterDurableSubscriptionsForNodeID(boolean isCoordinator
...
+            }
+            return count;
+
+        }
+    }
+
+    /**
+     * Gauge will return total number of topic subscriptions current node
+     */
+    private class TopicSubscriberGauge implements org.wso2.carbon.metrics.manager.Gauge {
```
Comment (pamod) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionStore.java]: Should we have this imported just like in the case of QueueSubscriberGauge ?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/distruptor/delivery/DeliveryEventHandler.java (hunk 2 above)
Comment (AnujaLK) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/distruptor/delivery/DeliveryEventHandler.java]: "this.getClass() + " part was removed by another PR.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/distruptor/inbound/MessagePreProcessor.java (hunk 3 above)
Comment (AnujaLK) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/distruptor/inbound/MessagePreProcessor.java]: "this.getClass() + " part was removed by another PR.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/store/rdbms/RDBMSMessageStoreImpl.java (hunk 4 above)
Comment (AnujaLK) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/store/rdbms/RDBMSMessageStoreImpl.java]: Fixed with PR https://github.com/wso2/andes/pull/276

File: modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionStore.java (hunk 5 above)
Comment (AnujaLK) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionStore.java]: Fixed with PR https://github.com/wso2/andes/pull/276

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/FlowControlManager.java [hunk 6]
```diff
@@ -25,16 +25,18 @@
...
 import org.wso2.andes.store.FailureObservingStoreManager;
 import org.wso2.andes.store.HealthAwareStore;
 import org.wso2.andes.store.StoreHealthListener;
-
 import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import org.wso2.carbon.metrics.manager.Gauge;
+import org.wso2.carbon.metrics.manager.Level;
+import org.wso2.carbon.metrics.manager.MetricManager;
 
 public class FlowControlManager  implements StoreHealthListener {
```
Comment (lafernando) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/FlowControlManager.java]: Add class level comments.

Comment (AnujaLK) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/FlowControlManager.java]: Fixed with PR https://github.com/wso2/andes/pull/276/files#diff-2016b84c6e6b1f4f3efb534920fe5eb0

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/Andes.java [hunk 7]
```diff
@@ -165,6 +169,10 @@ public void startSafeZoneAnalysisWorker() {
      */
     public void messageReceived(AndesMessage message, AndesChannel andesChannel, PubAckHandler pubAckHandler) {
         inboundEventManager.messageReceived(message, andesChannel, pubAckHandler);
+
+        //Adding metrics meter for message rate
+        Meter messageMeter = MetricManager.meter(Level.INFO, this.getClass() + MetricsConstants.MSG_RECEIVE_RATE);
```
Comment (imesh) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/Andes.java]: What would be the reason for passing the log level here?

Comment (AnujaLK) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/Andes.java]: Carbon Metrics API associate a Level with each Metric. These levels are similar to the Levels used in Logging Libraries. When we add a metric to metric manager, we need to provide level for that particular metric. Later we can set the level which we need to enable metrics.

**Pull request merged**